### DISCUSSION TITLE
Steps are not executed correctly when being ran in image built in previous job (from Github Packages)

### DISCUSSION THREAD

**Discussion by stex** [community/community#24960] [Actions]
<p>Hey!</p>
<p>It’s a bit difficult to find a good title for this issue (which might be the reason I didn’t find any solution for my problem through Google yet…)</p>
<p>I’m building a very simple image in the first job of my workflow:</p>
<pre><code class="lang-auto">FROM ruby:3

WORKDIR /app

COPY ./Gemfile ./Gemfile.lock ./
RUN bundle config set --local path '/gems'
RUN bundle install --jobs=$(nproc)

COPY ./ ./

CMD ["/bin/bash"]
</code></pre>
<p>This image is pushed to Github packages and I am able to pull it locally and run commands inside the corresponding container as expected, in this case</p>
<pre><code class="lang-bash">docker run -it --rm THE_IMAGE:TAG bundle exec standardrb
</code></pre>
<p>However, when running the job in Github Actions, I get the following result:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/f/2/f275f6b5dc29d720f9a75d99902190afa8d42b19.png" data-download-href="/uploads/short-url/yAUhBs8M7OP0FCJWd9wvh1Bl0wh.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/27221/181087893-b5c7f14f-f943-4037-9e65-6fb88907b78f.png" alt="image" data-base62-sha1="yAUhBs8M7OP0FCJWd9wvh1Bl0wh" width="434" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/f/2/f275f6b5dc29d720f9a75d99902190afa8d42b19_2_434x500.png, https://aws1.discourse-cdn.com/github/original/3X/f/2/f275f6b5dc29d720f9a75d99902190afa8d42b19.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/f/2/f275f6b5dc29d720f9a75d99902190afa8d42b19.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/f/2/f275f6b5dc29d720f9a75d99902190afa8d42b19_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">553×636 37.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I added a simple <code>ls -la</code> to see where I was. Surprisingly, the workspace inside the action seems to be empty, leading me to the assumption that I’m not inside the correct container at all.</p>
<p>My workflow looks like this:</p>
<pre><code class="lang-auto">name: test

on:
  - push
  - pull_request

env:
  REGISTRY: docker.pkg.github.com
  IMAGE_TAG: docker.pkg.github.com/stex/my_repo/the_image_name:${{ github.sha }}

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: satackey/action-docker-layer-caching@v0.0.11
        continue-on-error: true
      - run: docker login ${REGISTRY} -u ${GITHUB_ACTOR} -p ${{ github.token }}
      - run: docker build -t ${IMAGE_TAG} -f Dockerfile.dev .
      - run: docker push $IMAGE_TAG
  standardrb:
    runs-on: ubuntu-latest
    needs: build
    container:
      image: docker.pkg.github.com/stex/my_repo/the_image_name:${{ github.sha }}
      options: "--entrypoint /bin/bash"
      credentials:
        username: ${{ github.actor }}
        password: ${{ github.token }}
    steps:
      - run: ls -la
      - run: bundle exec standardrb
</code></pre>
<p>The documentation for running code inside an own container isn’t too extensive, but I’ve used this flow in Gitlab CI for a long time with great results.</p>
<p>Am I missing something really simple here?</p>
<p>Thanks a lot in advance!</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Your Dockerfile sets this:</p>
<aside class="quote no-group" data-username="Stex" data-post="1" data-topic="164026">
<div class="title">
<div class="quote-controls"></div>
 Stex:</div>
<blockquote>
<p><code>WORKDIR /app</code></p>
</blockquote>
</aside>
<p>So <code>/app</code> is where your data ends up. However, when you use a container image for a job (as in your second job) settings like working directory or entrypoint are ignored. The actions runner mounts the job workspace into the container and uses that as the working directory. The logs for the “Initialize containers” step should show you all the Docker arguments that are used to start the container.</p>
<p>If you want to run the things in <code>/app</code> within the container you’ll need to change the directory, either as part of your step or <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsrun">using the <code>working-directory</code> keyword</a>.</p>

**Comment by stex** [comment]
<p>Looks like I’ve been too spoiled by Gitlab where everything just works as expected (and every command is ran inside the same container).</p>
<pre><code class="lang-yaml">- run: cd /app &amp;&amp; bundle exec standardrb
</code></pre>
<p>solved the problem for me.</p>
<aside class="quote no-group" data-username="airtower-luna" data-post="2" data-topic="164026">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The logs for the “Initialize containers” step should show you all the Docker arguments that are used to start the container.</p>
</blockquote>
</aside>
<p>You are right, it does. I missed that <code>__w</code> is mounted in the container with <code>workdir</code> being set to a path inside it. In my defense, that’s a 5 line <code>docker create</code> statement :see_no_evil:</p>
<p>Thanks a lot for your help, maybe an actual example of how to properly use own, pre-built images in actions would make a good addition to the documentation.</p>
